### ISSUE TITLE
Response returned to wrong client

### ISSUE THREAD

**Issue by jesseops** [haproxy/haproxy#583]
## Output of `haproxy -vv` and `uname -a`

```
~# haproxy -vv
HA-Proxy version 2.0.13-2ppa1~bionic 2020/04/02 - https://haproxy.org/
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-ymCBCb/haproxy-2.0.13=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wno-implicit-fallthrough -Wno-stringop-overflow -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
Running on OpenSSL version : OpenSSL 1.1.1  11 Sep 2018
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.31 2018-02-12
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE     mux=H2
              h2 : mode=HTTP       side=FE        mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services :
	prometheus-exporter

Available filters :
	[SPOE] spoe
	[COMP] compression
	[CACHE] cache
	[TRACE] trace

:~# uname -a
Linux $FQDN 4.15.0-65-generic #74-Ubuntu SMP Tue Sep 17 17:06:04 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

## What's the configuration?

```
global
	description	$HAPROXY_HOSTNAME
	log /dev/log	local0
	chroot /var/lib/haproxy
	stats socket /run/haproxy/admin.sock mode 660 level admin expose-fd listeners # adding expose-fd and liseners since 2.0
	stats timeout 30s
	user haproxy
	group haproxy
	daemon

	maxconn 200000
	max-spread-checks 5000 #AR max time before haproxy comes up and will send out health checks to every server
	# maxsessrate <integer> #AR we should figure out a safe number for this and add it to protect the backend service
	# maxconnrate <integer> #AR we should figure out a safe number for this and add it to protect the backend service


	# Default SSL material locations
	ca-base /etc/ssl/certs
	crt-base /etc/tsheets/ssl

  	# Use Mozilla's recommended security settings
  	# See https://wiki.mozilla.org/Security/Server_Side_TLS#Intermediate_compatibility_.28recommended.29
  ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
        ssl-default-bind-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets

        ssl-default-server-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384
        ssl-default-server-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets

  	tune.ssl.default-dh-param 2048
  	# SSL session cache (20k by default). 200 bytes per entry, so 500k =~ 100MB AR: bumping to 1M
  	tune.ssl.cachesize 1000000

defaults
	log	global
	mode	http
  # Disabling HTX to rule out issues with JWT SUDO-607
  no option http-use-htx

	option	httplog
	option	dontlognull
        timeout connect 5000   # 5s timeout if backend is not found
        timeout client  50000  # 50s timeout for client
        timeout server  300000  # ma: Changed to 300s. 50s timeout for server originally
	errorfile 400 /etc/haproxy/errors/400.http
	errorfile 403 /etc/haproxy/errors/403.http
	errorfile 408 /etc/haproxy/errors/408.http
	errorfile 500 /etc/haproxy/errors/500.http
	errorfile 502 /etc/haproxy/errors/502.http
	errorfile 503 /etc/haproxy/errors/503.http
	errorfile 504 /etc/haproxy/errors/504.http


frontend wild_FE
        bind $VIRTUAL_IP:443 transparent ssl crt $CERTIFICATE
        maxconn 125000

        # Debug logging
        http-request capture req.hdr(Host) len 25
        http-request capture req.cook(SESSION_ID) len 25

	# JR: Added 04/2020 b/c HAProxy recommends being careful about POST methods
	http-request disable-l7-retry if METH_POST

        # Route requests with a 'SESSION_ID ...' cookie to the appropriate shard:
        # According to our source code, session IDs are formatted as S:<shard_id>$$<db_sesskey>
        # API code uses a different format in the Authorization header: Bearer S.<shard_id>__<db_sesskey>

        acl shard_1_sessid cook(SESSION_ID),url_dec -m beg S:1$$
        
        use_backend ajax-$SITEprivpub-s1_BE if shard_1_sessid

        # Finally, "catch-all" handles any remaining as-yet-unmatched traffic here:
        acl $SITE_dc_is_dead nbsrv($SITE-unswimlaned) lt 1
        use_backend $ALT_SITElb-wild-vip_BE if $SITE_dc_is_dead
        default_backend $SITE-unswimlaned


backend ajax-$SITEprivpub-s1_BE
        option forwardfor
        option log-health-checks
        option httpchk HEAD / HTTP/1.0
        http-check expect rstatus (2|3)[0-9][0-9]
        default-server inter 4s fall 3 rise 2
        balance roundrobin
        stick-table type integer size 2M expire 11d peers mypeers
        stick on req.cook(SESSION_ID),djb2
        stick store-request req.cook(SESSION_ID),djb2
        server $WEB_SERVER_HOSTNAME-privpub_FE   127.1.0.18:10000    source 127.1.0.18 ssl verify required ca-file /etc/ssl/certs/ca-certificates.crt maxconn 400 check weight 100
        server $ANOTHER_WEB_SERVER_HOSTNAME-privpub_FE   127.1.0.20:10000    source 127.1.0.20 ssl verify required ca-file /etc/ssl/certs/ca-certificates.crt maxconn 400 check weight 100


Example listener config:

listen $WEB_SERVER_HOSTNAME-privpub
        bind 127.1.0.72:10000
        mode tcp
        option tcplog
        option dontlog-normal

        server  $WEB_SERVER_HOSTNAME-int     $INTERNAL_IP:443   check
        server  $WEB_SERVER_HOSTNAME          $EXTERNAL_IP:443   check backup
```

## Steps to reproduce the behavior

1. Click login (bug causes ~3 simultaneous HTTP calls for the login)
2. 3 successful logins return, 2 are dropped by client (`CD--` or `CL--`) termination falgs
3. The login response for one of the 2 dropped logins is returned to a completely different client, whose HTTP request is never seen on a backend server

## Actual behavior

I have 2 clients, `client1` processing a login and `client2` making normal HTTP calls. We have Apache running mod_php as our backend. HAProxy is running `2.0.13` with HTX disabled, retry on POST methods disabled. We are using the default threading mode. The requests below are all associated with the same HAProxy thread ID. Each of the requests below are being routed to the same backend Apache web server.

`client1` issues 3 simultaneous  login requests (that’s a bug we have since fixed). One of these login requests succeeds, while the other two fail, generally with termination flags of CD--, but more rarely with CL--. `client1` continues along doing other work based on the logs I see in Apache. We have very few confirmed cases of this bug, so it's possible that only the cases with `CL--` trigger the bug.

`client2` makes an HTTP call at approximately the same time as `client1` performs the triple login. I can tell from the HAProxy logs that `client2` gets an HTTP response of the same size as the login response from `client1`, and I never see the backend receive the HTTP request from client2. Some seconds later, I see the IP previously associated with `client2` now using the the session id and transaction id associated with one of the “extra” login requests that failed.

My hypothesis is that when `client2` makes it’s HTTP request, HAProxy is returning the response from one of the login requests from `client1` which terminated early instead of routing the request through to the backend and getting a new response.

FWIW we disabled HTX previously due to some problems we had with headers being junked, I’m following up to see if we every fixed logging such that we could understand the failure mode better and test turning HTX back on

## Expected behavior

An HTTP request will not be returned to a different client under any circumstances.

## Do you have any idea what may have caused this?

It appears to be an issue where a session is terminated by the client while the response is in transit. The next HTTP call in line gets the response. 

## Do you have an idea how to solve the issue?

Not a clue, but I'm happy to provide any (sanitized) logs that may be useful and test possible reproduction scenarios.

**Comment by wtarreau** [comment]
Interesting case, thanks for reporting it.

Indeed, while there is really nothing that can allow some incoming data to be sent cross-client (since streams are end-to-end), it would be technically possible that with connection reuse we'd send a request over a server connection that was wrongly put into the idle list while it was still holding a pending response. That would obviously be a bug if that were the case.

Could you try to see if you can still trigger it with "http-reuse never" ?

If you cannot enable HTX, could you also try with a pre-HTX version (1.8) ? It could  be possible that during the conversion to HTX we messed up some error processing in the non-HTX version. The coexistence of the two was a real nightmare to deal with :-/

Thanks!

**Comment by jesseops** [comment]
Thanks for the feedback! We've added `http-reuse never` to our `defaults` section and will monitor for a few days. We've relied on customer reports so far to detect this and it's been a very rare occurrence.  I'll report back next week!

**Comment by bedis** [comment]
@jesseops do you have any update after changing `http-reuse` ?

**Comment by jesseops** [comment]
Just re-checked and we haven't received any more notices of this being triggered after disabling `http-reuse`.

**Comment by wtarreau** [comment]
Great, thanks for the update. Sadly this affects the old legacy mode which was dropped a year ago, so it will not be easy to debug this one :-/
We could imagine as a preventive measure to disable http-reuse by default when HTX is disabled but it's hiding a problem which might be deeper, I'd rather figure what's happening.

**Comment by wtarreau** [comment]
[ _i'm finding this comment in the text area after my browser died but can't seem to find it above while I was pretty sure I already responded - I hate web-based interfaces_ ]

Great, thanks for the update. Sadly this affects the old legacy mode which was dropped a year ago, so it will not be easy to debug this one :-/
We could imagine as a preventive measure to disable http-reuse by default when HTX is disabled but it's hiding a problem which might be deeper, I'd rather figure what's happening.

**Comment by wtarreau** [comment]
I forgot about this one which was reported differently as well and fixed in latest 2.0.x (the issue was in fact about what server was reused). Another 2.0 will be emitted very soon, please upgrade and confirm it's gone so that 1) you are safe and 2) we can close this once for all.

**Comment by lukastribus** [comment]
@jesseops any updates about this in 2.0.17 ?

**Comment by jesseops** [comment]
Hey thanks for checking back. We're in the process of migrating into AWS so changes to existing infrastructure are on hold; I should be able to test this update once that migration is complete as we still plan to use HAProxy for our routing layer.

**Comment by TimWolla** [comment]
@jesseops Checking in once again.

**Comment by jesseops** [comment]
Thanks Tim, we should be able to close this as I have moved on to a new company and I'm unaware of the current status of this testing back at my previous employer.

**Comment by TimWolla** [comment]
Thanks for your reply. Closing this issue then, as it's most likely fixed.
